pippy/PipelineSchedule.py

```diff
@@ -0,0 +1,460 @@
+                    dist.batch_isend_irecv(ops)
+
+                logger.info(
+                    f"{self._stage.stage_id} forward {i} finished, microbatch: {mb.shape}"
+                )
+
+        for i, _ in enumerate(microbatches):
+            with record_function(f"Backward {i}"):
+                ops = self._stage.get_bwd_recv_ops()
+                if ops:
+                    dist.batch_isend_irecv(ops).pop().wait()
+
+                self._stage.backward(
+                    is_first_mb=i == 0,
+                    is_last_mb=i == len(microbatches) - 1,
+                )
+
+                ops = self._stage.get_bwd_send_ops()
+                if ops:
+                    dist.batch_isend_irecv(ops)
+
+            logger.info(f"{self._stage.stage_id} backward {i} finished")
+
+
+class PipelineScheduleLoopedBFS:
+    def __init__(self, stages):
+        self._stages = stages
+
+    def step(self, microbatches):
+        for s, stage in enumerate(self._stages):
+            for i, mb in enumerate(microbatches):
+                with record_function(f"Stage {s} Forward"):
+                    is_last_mb = i == len(microbatches) - 1
+
+                    ops = stage.get_fwd_recv_ops()
+                    if ops:
+                        dist.batch_isend_irecv(ops).pop().wait()
+
+                    stage.forward(mb, is_first_mb=i == 0, is_last_mb=is_last_mb)
+
+                    ops = stage.get_fwd_send_ops()
+                    if ops:
+                        dist.batch_isend_irecv(ops)
+
+        for stage in reversed(self._stages):
+            for i in range(len(microbatches)):
+                with record_function(f"Stage {stage.stage_id} Backward"):
+                    ops = stage.get_bwd_recv_ops()
+                    if ops:
+                        dist.batch_isend_irecv(ops).pop().wait()
+
+                    stage.backward(
+                        is_first_mb=i == 0,
+                        is_last_mb=i == len(microbatches) - 1,
+                    )
+
+                    ops = stage.get_bwd_send_ops()
+                    if ops:
+                        dist.batch_isend_irecv(ops)
+
+
+class PipelineScheduleLoopedDFS:
+    def __init__(self, stages, n_microbatch, pp_id, n_pp):
+        assert (
+            n_microbatch % n_pp == 0
+        ), f"Looped DFS schedule requires microbatch_size ({n_microbatch}) to be a multiple of n_pp ({n_pp})"
+
+        self.stages = stages
+        self.n_microbatch = n_microbatch
+
+        self.n_stages = len(stages)
+        self.total_stages = self.n_stages * n_pp
+        # world_size
+        self.n_pp = n_pp
+
+        self.stage_id_to_global_stage_id = [
+            (i * n_pp) + pp_id for i in range(self.n_stages)
+        ]
+
+        # pp_id is the same as local rank within the PP dimension
+        self.pp_id = pp_id
+
+        # number of sequences (chunks) to divide microbatches into == microbatch_size / (microbatch_size / n_pp)
+        self.seq_size = n_pp
```

Comment:
I am a bit confused by this comment. What does seq_size refer to?

Reply:
the comment looks old, sequence size should be world_size